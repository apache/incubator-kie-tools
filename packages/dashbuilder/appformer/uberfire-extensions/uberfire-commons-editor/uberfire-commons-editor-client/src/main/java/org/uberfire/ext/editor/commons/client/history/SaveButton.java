/*
 * Copyright 2015 JBoss, by Red Hat, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.ext.editor.commons.client.history;

import com.google.gwt.core.client.GWT;
import com.google.gwt.event.shared.HandlerRegistration;
import com.google.gwt.user.client.ui.Widget;

import org.gwtbootstrap3.client.ui.AnchorListItem;
import org.gwtbootstrap3.client.ui.Button;
import org.gwtbootstrap3.client.ui.ButtonGroup;
import org.gwtbootstrap3.client.ui.DropDownMenu;
import org.gwtbootstrap3.client.ui.constants.ButtonSize;
import org.gwtbootstrap3.client.ui.constants.Toggle;
import org.uberfire.ext.editor.commons.client.resources.i18n.CommonConstants;
import org.uberfire.mvp.Command;
import org.uberfire.mvp.ParameterizedCommand;
import org.uberfire.workbench.model.menu.EnabledStateChangeListener;
import org.uberfire.workbench.model.menu.MenuCustom;
import org.uberfire.workbench.model.menu.MenuPosition;
import org.uberfire.workbench.model.menu.MenuVisitor;

public class SaveButton
        implements MenuCustom<Widget> {

    private ButtonGroup buttonGroup = GWT.create(ButtonGroup.class);
    private Button saveButton = GWT.create(Button.class);
    private DropDownMenu dropDownMenu = GWT.create(DropDownMenu.class);
    private Button caretButton = GWT.create(Button.class);
    private AnchorListItem saveWithCommentsButton = GWT.create(AnchorListItem.class);

    private HandlerRegistration saveButtonClickHandler;
    private HandlerRegistration saveWithCommentsButtonClickHandler;

    public SaveButton() {
        buttonGroup.add(saveButton);

        saveButton.setSize(ButtonSize.SMALL);
        saveButton.setText(CommonConstants.INSTANCE.Save());

        buttonGroup.add(caretButton);

        caretButton.setToggleCaret(true);
        caretButton.setDataToggle(Toggle.DROPDOWN);
        caretButton.setSize(ButtonSize.SMALL);
        caretButton.setMarginRight(10);

        buttonGroup.add(dropDownMenu);

        dropDownMenu.add(saveWithCommentsButton);

        saveWithCommentsButton.setText(CommonConstants.INSTANCE.SaveWithComments());
    }

    @Override
    public Widget build() {
        return buttonGroup;
    }

    @Override
    public boolean isEnabled() {
        return saveButton.isEnabled();
    }

    @Override
    public void setEnabled(boolean enabled) {
        saveButton.setEnabled(enabled);
        caretButton.setEnabled(enabled);
    }

    @Override
    public String getContributionPoint() {
        return null;
    }

    @Override
    public String getCaption() {
        return null;
    }

    @Override
    public MenuPosition getPosition() {
        return null;
    }

    @Override
    public int getOrder() {
        return 0;
    }

    @Override
    public void accept(final MenuVisitor visitor) {
        visitor.visit(this);
    }

    @Override
    public void addEnabledStateChangeListener(EnabledStateChangeListener listener) {

    }

    @Override
    public String getIdentifier() {
        return null;
    }

    public void setTextToSave() {
        saveButton.setText(CommonConstants.INSTANCE.Save());
        saveWithCommentsButton.setText(CommonConstants.INSTANCE.SaveWithComments());
    }

    public void setTextToRestore() {
        saveButton.setText(CommonConstants.INSTANCE.Restore());
        saveWithCommentsButton.setText(CommonConstants.INSTANCE.RestoreWithComments());
    }

    public void setCommand(Command command) {
        if (saveButtonClickHandler != null) {
            saveButtonClickHandler.removeHandler();
        }
        saveButtonClickHandler = saveButton.addClickHandler(e -> command.execute());

        if (saveWithCommentsButtonClickHandler != null) {
            saveWithCommentsButtonClickHandler.removeHandler();
        }
        caretButton.setVisible(false);
    }

    public void setCommand(ParameterizedCommand<Boolean> command) {
        if (saveButtonClickHandler != null) {
            saveButtonClickHandler.removeHandler();
        }
        saveButtonClickHandler = saveButton.addClickHandler(e -> command.execute(false));

        if (saveWithCommentsButtonClickHandler != null) {
            saveWithCommentsButtonClickHandler.removeHandler();
        }
        saveWithCommentsButtonClickHandler = saveWithCommentsButton.addClickHandler(e -> command.execute(true));
        caretButton.setVisible(true);
    }
}
