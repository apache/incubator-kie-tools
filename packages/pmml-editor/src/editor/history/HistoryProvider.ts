/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { applyPatches, produce } from "immer";
import type { WritableDraft } from "immer/dist/types/types-external";
import { PMML } from "@kogito-tooling/pmml-editor-marshaller/dist/marshaller/model";
import { cloneDeep, get, set } from "lodash";
import { Path } from "../paths";

interface Change {
  path: Path | null;
  change: any;
  reverse: any;
}

interface HistoryStore {
  changes: Change[];
  index: number;
}

interface BatchEntry<M> {
  state: M;
  path: Path | null;
  recipe: (draft: WritableDraft<M>) => void;
  validate?: (pmml: PMML) => void;
}

export type Listener = (id: string) => void;

export class HistoryService {
  private pending: Array<BatchEntry<any>> = new Array<BatchEntry<any>>();

  private readonly history: HistoryStore = {
    changes: [],
    index: 0,
  };

  constructor(private readonly listeners: Listener[]) {}

  public batch = <M>(
    state: M,
    path: Path | null,
    recipe: (draft: WritableDraft<M>) => void,
    validate?: (pmml: PMML) => void
  ): void => {
    this.pending.push({ state: state, path: path, recipe: recipe, validate: validate });
  };

  public commit = (state: PMML | undefined): PMML | undefined => {
    if (state === undefined) {
      return;
    }

    //Commit changes
    const newState = this.mutate(state, null, (draft) => {
      this.pending.forEach((be) => {
        const segment = be.path === null ? draft : get(draft, be.path.path);
        be.recipe(segment as WritableDraft<any>);
      });
    });

    //Validate changes
    this.pending.forEach((be) => {
      if (be.validate !== undefined) {
        be.validate(newState);
      }
    });

    //Signal commit to listeners
    this.listeners.forEach((listener) => listener(`Command${this.history.index}`));

    this.pending = new Array<BatchEntry<any>>();

    return newState;
  };

  private mutate = <M>(state: M, path: Path | null, recipe: (draft: WritableDraft<M>) => void) => {
    if (this.history.index < this.history.changes.length) {
      this.history.changes = this.history.changes.slice(0, this.history.index);
    }

    const newState: M = produce(state, recipe, (patches, inversePatches) => {
      this.history.changes.push({ path: path, change: patches, reverse: inversePatches });
      this.history.index = this.history.changes.length;
    });

    return newState;
  };

  public undo = (state: PMML): PMML => {
    if (this.history.index > 0) {
      const change: Change = this.history.changes[--this.history.index];
      return this.apply(state, change.path, change.reverse);
    }

    return state;
  };

  public redo = (state: PMML): PMML => {
    if (this.history.index < this.history.changes.length) {
      const change: Change = this.history.changes[this.history.index++];
      return this.apply(state, change.path, change.change);
    }

    return state;
  };

  public index = (): number => {
    return this.history.index;
  };

  public changes = (): Change[] => {
    return this.history.changes;
  };

  private apply = (state: PMML, path: Path | null, patch: any) => {
    if (path === null) {
      return applyPatches(state, patch);
    }
    const branch: any = get(state, path.path);
    const branchUndone: any = applyPatches(branch, patch);
    const newState: PMML = cloneDeep(state);
    return set(newState, path.path, branchUndone);
  };
}
