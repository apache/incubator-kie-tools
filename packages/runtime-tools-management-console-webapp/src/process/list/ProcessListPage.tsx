/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { useCallback, useEffect } from "react";
import { Card } from "@patternfly/react-core/dist/js/components/Card";
import { ProcessList } from "./ProcessList";
import { useRuntimeInfo, useRuntimeSpecificRoutes } from "../../runtime/RuntimeContext";
import { AuthSession, useAuthSessionsDispatch } from "../../authSessions";
import { useHistory } from "react-router";
import { useRoutes } from "../../navigation/Hooks";
import { useEnv } from "../../env/hooks/EnvContext";
import { useRuntimePageLayoutDispatch } from "../../runtime/RuntimePageLayoutContext";

export const ProcessListPage: React.FC = () => {
  const { env } = useEnv();
  const history = useHistory();
  const runtimeRoutes = useRuntimeSpecificRoutes();
  const routes = useRoutes();
  const { runtimeDisplayInfo } = useRuntimeInfo();
  const { setOnSelectAuthSession } = useAuthSessionsDispatch();
  const { setCurrentPageTitle, setBreadcrumbText, setBreadcrumbPath } = useRuntimePageLayoutDispatch();

  useEffect(() => {
    setOnSelectAuthSession(() => (authSession: AuthSession) => {
      history.push(runtimeRoutes.processes(authSession));
    });

    return () => {
      setOnSelectAuthSession(undefined);
    };
  }, [history, runtimeRoutes, setOnSelectAuthSession]);

  useEffect(() => {
    setCurrentPageTitle("Process Instances");
    setBreadcrumbText(["Home", runtimeDisplayInfo?.fullDisplayName ?? "Runtime", "Process Instances"]);
    setBreadcrumbPath([routes.home.path({}), runtimeRoutes.processes(), runtimeRoutes.processes()]);

    return () => {
      setCurrentPageTitle("");
      setBreadcrumbText([]);
      setBreadcrumbPath([]);
    };
  }, [
    routes.home,
    runtimeDisplayInfo?.fullDisplayName,
    runtimeRoutes,
    setBreadcrumbPath,
    setBreadcrumbText,
    setCurrentPageTitle,
  ]);

  const onNavigateToProcessDetails = useCallback(
    (processInstanceId: string) => {
      history.push(runtimeRoutes.processDetails(processInstanceId));
    },
    [history, runtimeRoutes]
  );

  useEffect(() => {
    document.title = `${env.RUNTIME_TOOLS_MANAGEMENT_CONSOLE_APP_NAME} :: Process Instances`;
  }, [env.RUNTIME_TOOLS_MANAGEMENT_CONSOLE_APP_NAME]);

  return (
    <Card className="kogito-management-console__card-size">
      <ProcessList onNavigateToProcessDetails={onNavigateToProcessDetails} />
    </Card>
  );
};
