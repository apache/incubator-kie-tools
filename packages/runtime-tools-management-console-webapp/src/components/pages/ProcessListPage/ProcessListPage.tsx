/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import * as React from "react";
import { useEffect } from "react";
import { Card } from "@patternfly/react-core/dist/js/components/Card";
import { PageSection } from "@patternfly/react-core/dist/js/components/Page";
import { RouteComponentProps } from "react-router-dom";
import ProcessListContainer from "../../containers/ProcessListContainer/ProcessListContainer";
import { StaticContext } from "react-router";
import * as H from "history";
import "../../styles.css";
import {
  OUIAProps,
  componentOuiaProps,
  ouiaPageTypeAndObjectId,
} from "@kie-tools/runtime-tools-components/dist/ouiaTools";
import { ProcessListState } from "@kie-tools/runtime-tools-process-gateway-api/dist/types";
import { PageSectionHeader } from "@kie-tools/runtime-tools-components/dist/components/PageSectionHeader";

interface MatchProps {
  instanceID: string;
}

const ProcessListPage: React.FC<RouteComponentProps<MatchProps, StaticContext, H.LocationState> & OUIAProps> = (
  props
) => {
  useEffect(() => {
    return ouiaPageTypeAndObjectId("process-instances");
  });
  const initialState: ProcessListState = props.location && (props.location.state as ProcessListState);

  return (
    <>
      <PageSectionHeader
        titleText="Process Instances"
        breadcrumbText={["Home", "Processes"]}
        breadcrumbPath={["/"]}
        ouiaId={props.ouiaId}
      />
      <PageSection {...componentOuiaProps(props.ouiaId, "page-section-content", props.ouiaSafe)}>
        <Card className="kogito-management-console__card-size">
          <ProcessListContainer initialState={initialState} />
        </Card>
      </PageSection>
    </>
  );
};

export default ProcessListPage;
