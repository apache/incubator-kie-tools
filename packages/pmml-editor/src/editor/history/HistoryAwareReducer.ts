/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Reducer } from "react";
import { HistoryService } from "./HistoryProvider";
import { ModelType } from "..";
import { Model } from "@kogito-tooling/pmml-editor-marshaller";

export type HistoryAwareReducer<S, A> = (service: HistoryService) => Reducer<S, A>;

export interface ModelReducerBinding<S, A> {
  reducer: Reducer<S, A>;
  //A Factory is required to instantiate an *Object* as opposed to just *JSON structure* as
  //instanceof is used to determine whether an editor exists for a specific Model type
  factory: (data: S) => S;
}

export type HistoryAwareModelReducer<A> = (
  service: HistoryService,
  modelReducers: Map<ModelType, ModelReducerBinding<any, any>>
) => Reducer<Model[], A>;
