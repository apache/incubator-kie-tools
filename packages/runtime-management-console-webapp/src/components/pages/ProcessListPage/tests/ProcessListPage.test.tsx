/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from "react";
import { mount } from "enzyme";
import ProcessListPage from "../ProcessListPage";
import { BrowserRouter } from "react-router-dom";
import * as H from "history";

jest.mock("../../../containers/ProcessListContainer/ProcessListContainer");

describe("ProcessListPage tests", () => {
  const props = {
    match: {
      params: {
        instanceID: "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
      },
      url: "",
      isExact: false,
      path: "",
    },
    location: H.createLocation(""),
    history: H.createBrowserHistory(),
  };
  it("Snapshot", () => {
    const wrapper = mount(
      <BrowserRouter>
        <ProcessListPage {...props} />
      </BrowserRouter>
    ).find("ProcessListPage");

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("MockedProcessListContainer").exists()).toBeTruthy();
  });
});
