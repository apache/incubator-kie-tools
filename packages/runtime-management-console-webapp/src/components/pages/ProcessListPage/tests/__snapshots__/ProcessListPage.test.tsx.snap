// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProcessListPage tests Snapshot 1`] = `
<ProcessListPage
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": false,
      "params": Object {
        "instanceID": "8035b580-6ae4-4aa8-9ec0-e18e19809e0b",
      },
      "path": "",
      "url": "",
    }
  }
>
  <PageSectionHeader
    breadcrumbPath={
      Array [
        "/",
      ]
    }
    breadcrumbText={
      Array [
        "Home",
        "Processes",
      ]
    }
    titleText="Process Instances"
  >
    <PageSection
      data-ouia-component-type="page-section-header"
      data-ouia-safe={true}
      variant="light"
    >
      <section
        className="pf-c-page__main-section pf-m-light"
        data-ouia-component-type="page-section-header"
        data-ouia-safe={true}
      >
        <Breadcrumb>
          <nav
            aria-label="Breadcrumb"
            className="pf-c-breadcrumb"
            data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
            data-ouia-component-type="PF4/Breadcrumb"
            data-ouia-safe={true}
          >
            <ol
              className="pf-c-breadcrumb__list"
            >
              <BreadcrumbItem
                key=".$0"
                showDivider={false}
              >
                <li
                  className="pf-c-breadcrumb__item"
                >
                  <Link
                    to="/"
                  >
                    <LinkAnchor
                      href="/"
                      navigate={[Function]}
                    >
                      <a
                        href="/"
                        onClick={[Function]}
                      >
                        Home
                      </a>
                    </LinkAnchor>
                  </Link>
                </li>
              </BreadcrumbItem>
              <BreadcrumbItem
                isActive={true}
                key=".$1"
                showDivider={true}
              >
                <li
                  className="pf-c-breadcrumb__item"
                >
                  <span
                    className="pf-c-breadcrumb__item-divider"
                  >
                    <AngleRightIcon
                      color="currentColor"
                      noVerticalAlign={false}
                      size="sm"
                    >
                      <svg
                        aria-hidden={true}
                        aria-labelledby={null}
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style={
                          Object {
                            "verticalAlign": "-0.125em",
                          }
                        }
                        viewBox="0 0 256 512"
                        width="1em"
                      >
                        <path
                          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                        />
                      </svg>
                    </AngleRightIcon>
                  </span>
                  Processes
                </li>
              </BreadcrumbItem>
            </ol>
          </nav>
        </Breadcrumb>
        <PageTitle
          title="Process Instances"
        >
          <Flex
            data-ouia-component-type="page-title"
            data-ouia-safe={true}
          >
            <div
              className="pf-l-flex"
              data-ouia-component-type="page-title"
              data-ouia-safe={true}
            >
              <FlexItem
                spacer={
                  Object {
                    "default": "spacerSm",
                  }
                }
              >
                <div
                  className="pf-m-spacer-sm"
                >
                  <Title
                    headingLevel="h1"
                    size="4xl"
                  >
                    <h1
                      className="pf-c-title pf-m-4xl"
                      data-ouia-component-id="OUIA-Generated-Title-1"
                      data-ouia-component-type="PF4/Title"
                      data-ouia-safe={true}
                    >
                      Process Instances
                    </h1>
                  </Title>
                </div>
              </FlexItem>
            </div>
          </Flex>
        </PageTitle>
      </section>
    </PageSection>
  </PageSectionHeader>
  <PageSection
    data-ouia-component-type="page-section-content"
    data-ouia-safe={true}
  >
    <section
      className="pf-c-page__main-section"
      data-ouia-component-type="page-section-content"
      data-ouia-safe={true}
    >
      <Card
        className="kogito-management-console__card-size"
      >
        <article
          className="pf-c-card kogito-management-console__card-size"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe={true}
          id=""
        >
          <MockedProcessListContainer />
        </article>
      </Card>
    </section>
  </PageSection>
</ProcessListPage>
`;
